### PR TITLE
NSX-T Edgegateway L2 VPN Tunnel session management

### PR DESCRIPTION
**Pull request by adezxc**
This PR adds support for L2 VPN Tunnel sessions

* Added new methods for `NsxtEdgeGateway`: `CreateL2VpnTunnel`, `GetAllL2VpnTunnels`, `GetL2VpnTunnelByName`, `GetL2VpnTunnelById`
* Added new types `types.EdgeL2VpnStretchedNetwork`, `types.EdgeL2VpnTunnelStatistics`, `types.EdgeL2VpnTunnelStatus`.
* Add new type `NsxtL2VpnTunnel` and its methods: `Refresh`, `Update`, `Statistics`, `Status`, `Delete`

## Place in UI
image

### CODE REVIEW

**Review by lvirbalas — APPROVED** [review]
I like the test! :)

File: govcd/nsxt_l2_vpn_tunnel.go [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+	"github.com/vmware/go-vcloud-director/v2/types/v56"
+)
+
+// NsxtL2VpnTunnel extend your organization VDC by enabling virtual machines to
+// maintain their network connectivity across geographical boundaries while keeping
+// the same IP address. The connection is secured with a route-based IPSec tunnel between the two sides of the tunnel.
+// You can configure the L2 VPN service on an NSX-T edge gateway in your VMware Cloud Director environment
+// and create a L2 VPN tunnel. Virtual machines remain on the same subnet, which enables you to extend
+// your organization VDC by stretching its network. This way, an edge gateway at one site can provide
+// all services to virtual machines on the other site.
```
Comment (adambarreiro) [inline govcd/nsxt_l2_vpn_tunnel.go:16]: Imho it looks better when not using 2nd person, but just a nitpick
```suggestion
// NsxtL2VpnTunnel extends an organization VDC by enabling virtual machines to
// maintain their network connectivity across geographical boundaries while keeping
// the same IP address. The connection is secured with a route-based IPSec tunnel between the two sides of the tunnel.
// The L2 VPN service can be configured on an NSX-T edge gateway in a VMware Cloud Director environment
// to create a L2 VPN tunnel. Virtual machines remain on the same subnet, which extends
// the organization VDC by stretching its network. This way, an edge gateway at one site can provide
// all services to virtual machines on the other site.
```

File: govcd/nsxt_l2_vpn_tunnel.go [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+
+	newTunnel, err := egw.GetL2VpnTunnelById(task.Task.Details)
+	if err != nil {
+		return nil, fmt.Errorf("error getting L2 VPN tunnel with id %s: %s", task.Task.Details, err)
+	}
+
+	return newTunnel, nil
+}
+
+func (l2Vpn *NsxtL2VpnTunnel) Refresh() error {
```
Comment (adambarreiro) [inline govcd/nsxt_l2_vpn_tunnel.go:63]: Missing docs

File: govcd/nsxt_l2_vpn_tunnel.go [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	err = client.OpenApiGetAllItems(apiVersion, urlRef, queryParameters, &typeResponses, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// Wrap all typeResponses into IpSpaceOrgAssignment types with client
```
Comment (adambarreiro) [inline govcd/nsxt_l2_vpn_tunnel.go:110]:
```suggestion
	// Wrap all typeResponses into NsxtL2VpnTunnel types with client
```

File: govcd/nsxt_l2_vpn_tunnel.go [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+	err = client.OpenApiPutItem(apiVersion, urlRef, nil, tunnelParams, &newTunnel.NsxtL2VpnTunnel, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return newTunnel, nil
+}
+
+// Delete deletes the L2 VPN Tunnel
+// On versions up to 10.5.0 (as of writing) there is a bug wtih deleting
```
Comment (adambarreiro) [inline govcd/nsxt_l2_vpn_tunnel.go:257]:
```suggestion
// On versions up to 10.5.0 (as of writing) there is a bug with deleting
```

File: govcd/nsxt_l2_vpn_tunnel_test.go [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+	check.Assert(serverTunnel.NsxtL2VpnTunnel.SessionMode, Equals, "SERVER")
+	check.Assert(serverTunnel.NsxtL2VpnTunnel.Enabled, Equals, true)
+	check.Assert(serverTunnel.NsxtL2VpnTunnel.LocalEndpointIp, Equals, localEndpointIp[0].IPAddress)
+	check.Assert(serverTunnel.NsxtL2VpnTunnel.RemoteEndpointIp, Equals, "1.1.1.1")
+	check.Assert(serverTunnel.NsxtL2VpnTunnel.ConnectorInitiationMode, Equals, "ON_DEMAND")
+	check.Assert(serverTunnel.NsxtL2VpnTunnel.PreSharedKey, Equals, check.TestName())
+
+	fetchedServerTunnel, err := edge.GetL2VpnTunnelById(serverTunnel.NsxtL2VpnTunnel.ID)
+	check.Assert(err, IsNil)
+	check.Assert(fetchedServerTunnel, DeepEquals, serverTunnel)
```
Comment (adambarreiro) [inline govcd/nsxt_l2_vpn_tunnel_test.go:80]: So this means that `PeerCode` is always returned on GET, did I understood it right? (I thought it was only retrievable on creation)

**Review by adambarreiro — COMMENTED** [review]
Amazing feature, it is really nice to have the comments describing the current quirks, they make the whole thing more understandable.
I'll run the tests and approve afterwards, just sharing few little things I found and some doubts.

**Review by adambarreiro — APPROVED** [review]
Test passed for me on VCD 10.5.0

File: govcd/nsxt_l2_vpn_tunnel_test.go [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+
+	check.Assert(updatedClientTunnel.NsxtL2VpnTunnel.Name, Equals, check.TestName())
+	check.Assert(updatedClientTunnel.NsxtL2VpnTunnel.Description, Equals, check.TestName())
+	check.Assert(updatedClientTunnel.NsxtL2VpnTunnel.SessionMode, Equals, "CLIENT")
+	check.Assert(updatedClientTunnel.NsxtL2VpnTunnel.Enabled, Equals, true)
+	check.Assert(updatedClientTunnel.NsxtL2VpnTunnel.LocalEndpointIp, Equals, localEndpointIp[0].IPAddress)
+	check.Assert(updatedClientTunnel.NsxtL2VpnTunnel.RemoteEndpointIp, Equals, "2.2.2.2")
+
+	// There is a bug in all versions up to 10.5.0, it happens
+	// when a L2 VPN Tunnel is created in CLIENT mode, has atleast one Org VDC
```
Comment (dataclouder) [inline govcd/nsxt_l2_vpn_tunnel_test.go:165]:
```suggestion
	// when a L2 VPN Tunnel is created in CLIENT mode, has at least one Org VDC
```

File: govcd/nsxt_l2_vpn_tunnel_test.go (hunk 5 above)
Comment (adezxc) [reply govcd/nsxt_l2_vpn_tunnel_test.go:80]: There is a quirky behavior which should be fixed in the next version of API, but basically to summarize everything right now:

The server tunnel asks for pre-shared key on creation and always returns both pre-shared key and peer code on GET requests.
The client tunnel asks for peer code on creation and *shouldn't* return any of those two fields, but it returns only the decoded pre-shared key

File: govcd/nsxt_l2_vpn_tunnel.go [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+// and create a L2 VPN tunnel. Virtual machines remain on the same subnet, which enables you to extend
+// your organization VDC by stretching its network. This way, an edge gateway at one site can provide
+// all services to virtual machines on the other site.
+type NsxtL2VpnTunnel struct {
+	NsxtL2VpnTunnel *types.NsxtL2VpnTunnel
+	client          *Client
+	// edgeGatewayId is stored for usage in NsxtFirewall receiver functions
+	edgeGatewayId string
+}
+
```
[the remark this answers is not in the document]
Comment (adezxc) [reply govcd/nsxt_l2_vpn_tunnel.go:23]: I thought it's more of a personal code style choice, but I could implement it, it's just that some of other NSX-T resources didn't have it which is why I didn't pick this up.

File: govcd/nsxt_l2_vpn_tunnel.go [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+	newTunnel, err := egw.GetL2VpnTunnelById(task.Task.Details)
+	if err != nil {
+		return nil, fmt.Errorf("error getting L2 VPN tunnel with id %s: %s", task.Task.Details, err)
+	}
+
+	return newTunnel, nil
+}
+
+func (l2Vpn *NsxtL2VpnTunnel) Refresh() error {
+	client := l2Vpn.client
```
[the remark this answers is not in the document]
Comment (adezxc) [reply govcd/nsxt_l2_vpn_tunnel.go:64]: The latter suggestion could work, with a function like `getL2VpnTunnelByEdgeGatewayAndId(edgeGatewayId string, tunnelId string)` but I'm not sure if that would be the better choice, even though we're repeating a bit of code here

File: govcd/nsxt_l2_vpn_tunnel.go (hunk 1 above)
Comment (adezxc) [reply govcd/nsxt_l2_vpn_tunnel.go]: Thanks for the suggestion, changed

File: govcd/nsxt_l2_vpn_tunnel.go (hunk 2 above)
Comment (adezxc) [reply govcd/nsxt_l2_vpn_tunnel.go:64]: Added

File: govcd/nsxt_l2_vpn_tunnel.go (hunk 3 above)
Comment (adezxc) [reply govcd/nsxt_l2_vpn_tunnel.go]: Fixed

File: govcd/nsxt_l2_vpn_tunnel.go (hunk 4 above)
Comment (adezxc) [reply govcd/nsxt_l2_vpn_tunnel.go]: Thanks, fixed!

File: .changes/v2.22.0/619-features.md [hunk 9]
```diff
@@ -0,0 +1 @@
+* Added support for L2 VPN Tunnel session management via types `types.NsxtL2VpnTunnel`, `types.EdgeL2VpnStretchedNetwork`, `types.EdgeL2VpnTunnelStatistics`, `types.EdgeL2VpnTunnelStatus`. Methods for `NsxtEdgeGateway`: `CreateL2VpnTunnel`, `GetAllL2VpnTunnels`, `GetL2VpnTunnelByName`, `GetL2VpnTunnelById`, methods for `NsxtL2VpnTunnel`: `Refresh`, `Update`, `Statistics`, `Status`, `Delete` [GH-619]
```
[the remark this answers is not in the document]
Comment (adezxc) [reply .changes/v2.22.0/619-features.md]: Changed

File: govcd/nsxt_l2_vpn_tunnel.go [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+
+	tunnel, err := oneOrError("name", name, results)
+	if err != nil {
+		return nil, err
+	}
+
+	return tunnel, nil
+}
+
+// GetL2VpnTunnelById gets the L2 VPN Tunnel by its' ID
```
[the remark this answers is not in the document]
Comment (adezxc) [reply govcd/nsxt_l2_vpn_tunnel.go]: Fixed, for some reason I thought `its` and `its'` can be used interchangeably

File: govcd/nsxt_l2_vpn_tunnel.go [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+func (l2Vpn *NsxtL2VpnTunnel) Update(tunnelParams *types.NsxtL2VpnTunnel) (*NsxtL2VpnTunnel, error) {
+	if l2Vpn.NsxtL2VpnTunnel.SessionMode != tunnelParams.SessionMode {
+		return nil, fmt.Errorf("error updating the L2 VPN Tunnel: session mode can't be changed after creation")
+	}
+
+	if tunnelParams.SessionMode == "CLIENT" && !tunnelParams.Enabled {
+		// There is a known bug up to 10.5.0, the CLIENT sessions can't be
+		// disabled and can result in unexpected behaviour for the following
+		// operations
+		return nil, fmt.Errorf("client sessions can't be disabled")
```
[the remark this answers is not in the document]
Comment (adezxc) [reply govcd/nsxt_l2_vpn_tunnel.go]: A constraint won't hurt, added. Though I think the resolution might be that a CLIENT session won't have a state, but let's see.

File: govcd/nsxt_l2_vpn_tunnel_test.go (hunk 6 above)
Comment (adezxc) [reply govcd/nsxt_l2_vpn_tunnel_test.go]: Fixed

File: govcd/nsxt_l2_vpn_tunnel.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (Didainius) [reply govcd/nsxt_l2_vpn_tunnel.go:65]: We do the thing that Giuseppe mentioned quite often, but it is up to you

**Review by Didainius — CHANGES_REQUESTED** [review]
Thanks - a few queries from me

File: govcd/nsxt_l2_vpn_tunnel.go [hunk 12]
```diff
@@ -0,0 +1,284 @@
...
+		}
+	}
+
+	return results, nil
+}
+
+// GetL2VpnTunnelByName gets the L2 VPN Tunnel by name
+func (egw *NsxtEdgeGateway) GetL2VpnTunnelByName(name string) (*NsxtL2VpnTunnel, error) {
+	queryParams := url.Values{}
+	queryParams.Add("name", name)
```
[the remark this answers is not in the document]
Comment (adezxc) [reply govcd/nsxt_l2_vpn_tunnel.go]: Ah, thanks for spotting, it's a case that I missed, the `GET
https://{api_host}/cloudapi/1.0.0/edgeGateways/{gatewayId}/l2vpn/tunnels` endpoint doesn't support filtering, will fix the function and add a test case :)

File: govcd/nsxt_l2_vpn_tunnel.go [hunk 13]
```diff
@@ -0,0 +1,284 @@
...
+func (l2Vpn *NsxtL2VpnTunnel) Update(tunnelParams *types.NsxtL2VpnTunnel) (*NsxtL2VpnTunnel, error) {
+	if l2Vpn.NsxtL2VpnTunnel.SessionMode != tunnelParams.SessionMode {
+		return nil, fmt.Errorf("error updating the L2 VPN Tunnel: session mode can't be changed after creation")
+	}
+
+	if tunnelParams.SessionMode == "CLIENT" && !tunnelParams.Enabled {
+		// There is a known bug up to 10.5.0, the CLIENT sessions can't be
+		// disabled and can result in unexpected behaviour for the following
+		// operations
+		if l2Vpn.client.APIVCDMaxVersionIs("38.0") {
```
[the remark this answers is not in the document]
Comment (adezxc) [reply govcd/nsxt_l2_vpn_tunnel.go]: Updated, thanks for spotting. I added a step with setting the `Enabled` field for CLIENT tunnels and API versions above `38.0` so when we test it on 10.5.1 we can see the results.

Also added a similar check to the `de-attach networks` step before deletion

File: govcd/nsxt_l2_vpn_tunnel_test.go [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+)
+
+// Test_NsxtL2VpnTunnel tests NSX-T Edge Gateway L2 VPN Tunnels.
+// 1. It creates/gets/updates/deletes a SERVER type Tunnel, also the peer code (encoded configuration of the tunnel)
+// is saved for creation of CLIENT type tunnel.
+// 2. Creates/gets/updates/deletes a CLIENT type Tunnel
+func (vcd *TestVCD) Test_NsxtL2VpnTunnel(check *C) {
+	skipNoNsxtConfiguration(vcd, check)
+	skipOpenApiEndpointTest(vcd, check, types.OpenApiPathVersion1_0_0+types.OpenApiEndpointEdgeGatewayL2VpnTunnel)
+
```
[the remark this answers is not in the document]
Comment (adezxc) [reply govcd/nsxt_l2_vpn_tunnel_test.go:20]: Ah, correct, fixed.

File: govcd/nsxt_l2_vpn_tunnel.go [hunk 15]
```diff
@@ -0,0 +1,284 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	urlRef, err := client.OpenApiBuildEndpoint(fmt.Sprintf(endpoint, egw.EdgeGateway.ID), id)
+	if err != nil {
+		return nil, err
+	}
+
+	updatedTunnel := &NsxtL2VpnTunnel{
```
[the remark this answers is not in the document]
Comment (adezxc) [reply govcd/nsxt_l2_vpn_tunnel.go]: Thanks for spotting, fixed.

File: govcd/nsxt_l2_vpn_tunnel.go [hunk 16]
```diff
@@ -0,0 +1,282 @@
...
+	if l2Vpn.NsxtL2VpnTunnel.SessionMode != tunnelParams.SessionMode {
+		return nil, fmt.Errorf("error updating the L2 VPN Tunnel: session mode can't be changed after creation")
+	}
+
+	if tunnelParams.SessionMode == "CLIENT" && !tunnelParams.Enabled {
+		// There is a known bug up to 10.5.0, the CLIENT sessions can't be
+		// disabled and can result in unexpected behaviour for the following
+		// operations
+		if l2Vpn.client.APIVCDMaxVersionIs("<= 38.0") {
+			return nil, fmt.Errorf("client sessions can't be disabled")
```
Comment (lvirbalas) [inline govcd/nsxt_l2_vpn_tunnel.go:227]: What about extending the error message to describe why (prior to which version) it can't be disabled?

**Review by dataclouder — APPROVED** [review]

**Review by Didainius — APPROVED** [review]

File: govcd/nsxt_l2_vpn_tunnel.go (hunk 16 above)
Comment (adezxc) [reply govcd/nsxt_l2_vpn_tunnel.go:228]: I think I'll leave it, it's not certain if this will be allowed in future versions

Comment (adezxc) [reply govcd/nsxt_l2_vpn_tunnel.go:228]: Done, but I'm not sure if they add the option in future versions as it might be intended

**Pull request merged**